### ISSUE TITLE
[java] DU anomaly false positive?

### ISSUE THREAD

**Issue by MPoorter** [pmd/pmd#1304]
**Affects PMD Version:** 6.1.0

**Rule:** [DataflowAnomalyAnalysis](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#dataflowanomalyanalysis)

**Description:**

> DU - Anomaly: A recently defined variable is undefined. These anomalies may appear in normal source text.

I don't think this is the case, as the trimmedValue is used in the stream operation?

**Code Sample demonstrating the issue:**

``` java
import static org.apache.commons.lang3.StringUtils.trimToNull;

public class DummyService {

    public boolean dummyMethod(final String stringValue, final Set<String> dummySet) {
        final String trimmedValue = trimToNull(stringValue);
        return dummySet.stream()
            .noneMatch(value -> value.equalsIgnoreCase(trimmedValue));
    }

}
```

Result:
`n/a:6:	Found 'DU'-anomaly for variable 'trimmedValue' (lines '6'-'9').`

**Running PMD through:** *CLI*

**Comment by jsotuyod** [comment]
@MPoorter thanks for the report!

Unfortunately, this rule is by far the poorest in PMD. The data flow codebase has long been outdated, and has little or no understanding of any Java construct introduced in Java 5 or later (for-each-loops, try-with-resources, lambdas...), so these kind of false positives are not uncommon.

We want to revamp it at some point, but so far such an effort is unscheduled. It may be a good GSoC project candidate for 2019.

**Comment by adangel** [comment]
The rule [DataflowAnomalyAnalysis](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#dataflowanomalyanalysis) has been deprecated in favour of [UnusedAssignment](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedassignment), which is available with PMD 6.26.0.

Therefore this issue wont be fixed.

See also #2647.
